doc/developer/design/20230814_unbilled_replicas.md

```diff
@@ -0,0 +1,257 @@
+notices that after a version upgrade, Materialize needs additional compute
+resources, support should be able to add such resources for a limited amount of
+time without requiring explicit approval and complicated reimbursement for the
+customer.
+
+We extend Materialize with a concept of unbilled replicas, which server compute
+traffic but are not charged to a customer's account.
+
+## Alternative 1: Replica attributes
+
+We extend the definition of a replica to include a `BILLED AS` attribute.
+It determines the size of a replica for billing purposes, which can be different to the actual size, and includes a "free" option that is charged at zero credits.
+This attribute can only be specified on replicas, but not as an option for managed clusters.
+Access is limited to privileged accounts.
+To achieve this, we need to adjust managed clusters and billing logic.
+
+### Managed clusters
+
+At the moment, a managed cluster only contains replicas that conform to the spec of the cluster, but it cannot have replicas outside its definition.
+We change managed clusters to allow both the set of managed replicas, and additional replicas that are not covered by the configuration of the cluster.
+We achieve this by adding a `internal` flag to replicas to differentiate those that are part of the managed configuration and those that are manually configured, and adjust the logic around managed cluster to take this flag into account.
+
+Additional replicas that are not managed must not use names that might collide with the names of managed replicas.
+For this reason, they cannot be named `rX` where `X` is any number.
+
+### Billing
+
+Billing calculates credit consumption based on entries in the audit log.
+The query needs to take to use the size of a replica into account, unless a replicas has a specific billing property.
+
+### Detailed description
+
+We extend the syntax to create replicas with a `BILLED AS` and `INTERNAL` option:
+
+```sql
+CREATE CLUSTER REPLICA clsname.unbilled_replica SIZE 'xlarge', BILLED AS 'free', INTERNAL;
+
+CREATE CLUSTER REPLICA clsname.r2 SIZE 'xlarge', BILLED AS 'free', INTERNAL;
+ERROR: The cluster name 'r2' is reserved for managed replicas.
+```
+
+The properties mean:
+* `BILLED AS` determines the size at which a replica should be billed instead of its actual size,
+* `INTERNAL` indicates that a replica is created by Materialize and should never be part of a managed cluster.
+
+This syntax is restricted to privileged users, such as Materialize's support user, and cannot be accessed by the customer.
+
+When an unprivileged user tries to create the billing property, they'll receive an error:
+
+```sql
+CREATE CLUSTER REPLICA clsname.unbilled_replica SIZE 'xlarge', BILLED AS 'free';
+ERROR: BILLED AS and INTERNAL not permitted for non-system users.
+HINT: Contact support for billing-related inquiries.
+```
+
+We introduce an internal `mz_managed_cluster_replicas` table that lists replicas managed by managed clusters.
```

Comment:
Ah, now that I see the whole design together, would it be better to instead have the table `mz_internal.mz_internal_cluster_replicas`? Seems potentially more future proof, as we'd like to delete the `MANAGED` flag soon but likely to keep the `INTERNAL` flag around much longer.